hal/src/main/native/sim/DriverStation.cpp

```diff
@@ -90,6 +91,29 @@ static std::atomic<JoystickDataCache*> currentCache{nullptr};
 static JoystickDataCache* lastGiven = &caches[1];
 static JoystickDataCache* cacheToUpdate = &caches[2];
 
+namespace {
+struct TcpCache {
+  TcpCache() { std::memset(this, 0, sizeof(*this)); }
+  void Update();
+  void CloneTo(TcpCache* other) { std::memcpy(other, this, sizeof(*this)); }
+
+  HAL_MatchInfo matchInfo;
+  HAL_JoystickDescriptor descriptors[HAL_kMaxJoysticks];
+};
+static_assert(std::is_standard_layout_v<TcpCache>);
+}  // namespace
+
+static TcpCache tcpCache;
+static TcpCache tcpCurrent;
+
+void TcpCache::Update() {
+  SimDriverStationData->GetMatchInfo(&matchInfo);
+
+  for (int i = 0; i < HAL_kMaxJoysticks; i++) {
+    SimDriverStationData->GetJoystickDescriptor(i, &descriptors[i]);
+  }
+}
```

Comment:
Why is this function not inlined within the struct while the others are?

Reply:
Just copied from the roborio implementation. No particular reason otherwise.